### PR TITLE
Fix #95 - Make Geometry Preview work on Ubuntu 20.04 (workaround)

### LINKED ISSUE #95: Geometry Editor not working on Ubuntu
Issue overview
--------------

```
$ /usr/local/openstudioapplication-1.0.0-rc1/bin/OpenStudioApp
[0407/123432.277030:WARNING:resource_bundle_qt.cpp(115)] locale_file_path.empty() for locale 
Qt WebEngine ICU data not found at /home/dmacumbe/OpenStudio/super-build/Qt-prefix/src/Qt-install/resources. Trying parent directory...
Qt WebEngine ICU data not found at /home/dmacumbe/OpenStudio/super-build/Qt-prefix/src/Qt-install. Trying application directory...
Installed Qt WebEngine locales directory not found at location /home/dmacumbe/OpenStudio/super-build/Qt-prefix/src/Qt-install/translations/qtwebengine_locales. Trying application directory...
Qt WebEngine locales directory not found at location /usr/local/openstudioapplication-1.0.0-rc1/bin/qtwebengine_locales. Trying fallback directory... Translations MAY NOT not be correct.
Qt WebEngine resources not found at /home/dmacumbe/OpenStudio/super-build/Qt-prefix/src/Qt-install/resources. Trying parent directory...
Qt WebEngine resources not found at /home/dmacumbe/OpenStudio/super-build/Qt-prefix/src/Qt-install. Trying application directory...
[0407/123432.320588:WARNING:resource_bundle_qt.cpp(115)] locale_file_path.empty() for locale 

DevTools listening on ws://127.0.0.1:41993/devtools/browser/f7532e07-6bd4-4ade-b6a4-03e91f8b029c
Critical: THREE.WebGLRenderer: Error creating WebGL context.
Critical: Uncaught TypeError: Cannot read property 'getShaderPrecisionFormat' of null

```

**Note**: `cd /usr/local/openstudioapplication-1.0.0-rc1/bin/ && ./OpenStudioApp` does the same.

## Current Behavior

Doesn't work

## Expected Behavior

Should work

## Steps to Reproduce

1. Load the OSApp
2. Go to the "Geometry Editor" tab

### Environment

Some additional details about your environment for this issue (if relevant):
 - Platform (Operating system, version): Ubuntu 18.04
 - Version of OpenStudioApplication (if using an intermediate build, include SHA): 1.0.0-rc1, 0845fd16bb

### PR DESCRIPTION
**Pull request by jmarrec**
Fix #95

It took me 6 hours to find a workaround to make it work... I have an open stack overflow question here: https://stackoverflow.com/questions/67692525/cannot-load-qrc-html-file-in-qwebengineview-on-ubuntu-20-04-only

For now, we'll just read the entire content of the HTML file, then call QWebEngineView::setHtml ...

What's driving me insane is that it works with GeometryEditorView: https://github.com/openstudiocoalition/OpenStudioApplication/blob/e47cbb1e9ca01179fe5547f96100291392902b5f/src/openstudio_lib/GeometryEditorView.cpp#L240

**Edit**: Bumped ThreeJS version instead

### CODE REVIEW

File: src/openstudio_lib/GeometryPreviewView.cpp [hunk 1]
```diff
@@ -102,27 +100,74 @@ PreviewWebView::PreviewWebView(bool isIP, const model::Model& model, QWidget* t_
...
   m_view = new QWebEngineView(this);
-  m_view->settings()->setAttribute(QWebEngineSettings::WebAttribute::LocalContentCanAccessRemoteUrls, true);
-  m_view->settings()->setAttribute(QWebEngineSettings::WebAttribute::SpatialNavigationEnabled, true);
-
   m_page = new OSWebEnginePage(this);
   m_view->setPage(m_page);  // note, view does not take ownership of page
 
+
   connect(m_view, &QWebEngineView::loadFinished, this, &PreviewWebView::onLoadFinished);
-  //connect(m_view, &QWebEngineView::loadProgress, this, &PreviewWebView::onLoadProgress);
```
Comment (macumber) [inline src/openstudio_lib/GeometryPreviewView.cpp:112]: These drive the progress bar when loading right?  That is useful for feedback when loading large models, can we keep these lines?

File: src/openstudio_lib/GeometryPreviewView.cpp [hunk 2]
```diff
@@ -102,27 +100,74 @@ PreviewWebView::PreviewWebView(bool isIP, const model::Model& model, QWidget* t_
...
+
+
   // Qt 5.8 and higher
   m_view->settings()->setAttribute(QWebEngineSettings::AllowRunningInsecureContent, true);
+
+  m_view->settings()->setAttribute(QWebEngineSettings::WebAttribute::LocalContentCanAccessRemoteUrls, true);
+  m_view->settings()->setAttribute(QWebEngineSettings::WebAttribute::SpatialNavigationEnabled, true);
+
+  /*
+  m_view->settings()->setAttribute(QWebEngineSettings::JavascriptCanAccessClipboard, true);
```
Comment (macumber) [inline src/openstudio_lib/GeometryPreviewView.cpp:127]: You think these are unneeded?

File: src/openstudio_lib/GeometryPreviewView.cpp [hunk 3]
```diff
@@ -102,27 +100,74 @@ PreviewWebView::PreviewWebView(bool isIP, const model::Model& model, QWidget* t_
...
+  }
+
+  /*
+  // This works on Windows, macOS, Ubuntu1804, fails on Ubuntu 20.04
+  QUrl previewURL("qrc:/library/geometry_preview.html");
+  // QUrl previewURL("qrc://library/geometry_preview.html"); // => Warning: QResource '/geometry_preview.html' not found or is empty ((null):0, (null))
+  // QUrl previewURL("qrc:///library/geometry_preview.html");
+  m_view->load(previewURL);
+
+  // Local FileSystem: This works on Ubuntu 20.04 as well....
```
Comment (macumber) [inline src/openstudio_lib/GeometryPreviewView.cpp:156]: So maybe something is broken loading html from a QRC on Ubuntu 20.04?

**Comment by macumber** [comment]
Wow, thanks for digging into this @jmarrec, I will build and test on Ubuntu 20.04 and Windows.  It seems like some of the debugging stuff can get reverted but the overall approach seems to be ok (read QRC to string then call setHtml), seems like maybe there is a bug reading html from QRC?  It's possible we could make that a platform dependent switch?

File: src/openstudio_lib/GeometryPreviewView.cpp (hunk 2 above)
Comment (jmarrec) [reply src/openstudio_lib/GeometryPreviewView.cpp:127]: They weren't enabled before. During testing I enabled them, but it didn't make any difference

**Comment by macumber** [comment]
As noted on https://github.com/openstudiocoalition/OpenStudioApplication/issues/233 we should update to the newest ThreeJS at some point.   https://www.ladybug.tools/spider-2020/spider-idf-viewer/ should be a good example

**Comment by jmarrec** [comment]
On my ubuntu 20.04 I can see the maximize button jsut fine

image


When maximized too:

image

**Pull request merged**